### PR TITLE
Feature/45 create resumes

### PR DESCRIPTION
**Pull request by gustavo-rojas-jalafund**
Feature/45 Create Resumes

https://dev.azure.com/funjaladev/DevLevel3/_workitems/edit/45

### CODE REVIEW

**Review by ela-ocana-jalafund — DISMISSED** [review]

File: src/core/Jalasoft.TeamUp.Resumes.Core.Interfaces/IHealthService.cs [hunk 1]
```diff
@@ -5,5 +5,7 @@
     public interface IHealthService
     {
         Health GetHealth();
+
+        Health PostHealth();
```
Comment (rodrigo-baldiviezo-jalafund) [inline src/core/Jalasoft.TeamUp.Resumes.Core.Interfaces/IHealthService.cs]: it's not necesarry add a posthealth function

File: src/core/Jalasoft.TeamUp.Resumes.Core/HealthService.cs [hunk 2]
```diff
@@ -17,5 +17,10 @@ public Health GetHealth()
         {
             return this.healthRepository.GetHealth();
         }
+
+        public Health PostHealth()
+        {
+            return this.healthRepository.PostHealth();
+        }
```
Comment (rodrigo-baldiviezo-jalafund) [inline src/core/Jalasoft.TeamUp.Resumes.Core/HealthService.cs]: same, it's not necesarry posthealth

File: src/dal/Jalasoft.TeamUp.Resumes.DAL.Interfaces/IHealthRepository.cs [hunk 3]
```diff
@@ -5,5 +5,7 @@
     public interface IHealthRepository
     {
         Health GetHealth();
+
+        Health PostHealth();
```
Comment (rodrigo-baldiviezo-jalafund) [inline src/dal/Jalasoft.TeamUp.Resumes.DAL.Interfaces/IHealthRepository.cs]: delete this lines

File: src/dal/Jalasoft.TeamUp.Resumes.DAL/HealthRepository.cs [hunk 4]
```diff
@@ -9,5 +9,10 @@ public Health GetHealth()
         {
             return new Health { Message = "I'm resumes-api and I'm alive and running! ;)" };
         }
+
+        public Health PostHealth()
+        {
+            return new Health { Message = "I'm resumes-create-api and I'm alive and running! ;)" };
+        }
```
Comment (rodrigo-baldiviezo-jalafund) [inline src/dal/Jalasoft.TeamUp.Resumes.DAL/HealthRepository.cs]: delete this lines

File: src/dal/Jalasoft.TeamUp.Resumes.DAL/ResumesRepository.cs [hunk 5]
```diff
@@ -7,7 +7,7 @@
 
     public class ResumesRepository : IResumesRepository
     {
-        private static readonly Resume[] Resumes = new Resume[]
```
Comment (rodrigo-baldiviezo-jalafund) [inline src/dal/Jalasoft.TeamUp.Resumes.DAL/ResumesRepository.cs:11]: use this Resume[] to save your Resume

File: src/dal/Jalasoft.TeamUp.Resumes.DAL/ResumesRepository.cs [hunk 6]
```diff
@@ -81,7 +81,13 @@ public class ResumesRepository : IResumesRepository
...
         public IEnumerable<Resume> GetResumes()
         {
-            return Resumes;
+            return resumes;
+        }
+
+        public Resume PostResumes(Resume resume)
+        {
+            resumes.Add(resume);
+            return resume;
```
Comment (Raul-Gamarra-jalafund) [inline src/dal/Jalasoft.TeamUp.Resumes.DAL/ResumesRepository.cs:97]: in this case you should be returning just one resume not a list of resumes

File: tests/api/Jalasoft.TeamUp.Resumes.API.Tests/PostResumesTests.cs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        private readonly ResumesService resumeService;
+        private readonly Mock<IResumesRepository> mockRepository;
+
+        public PostResumesTests()
+        {
+            this.mockRepository = new Mock<IResumesRepository>();
+            this.resumeService = new ResumesService(this.mockRepository.Object);
+        }
+
+        [Fact]
```
Comment (Raul-Gamarra-jalafund) [inline tests/api/Jalasoft.TeamUp.Resumes.API.Tests/PostResumesTests.cs]: please add an Assert to verify the response code is the correct (201) one and also please create negative use test where the object could not be created and the response code is an error (adding validations in the service layer)

Comment (Raul-Gamarra-jalafund) [reply tests/api/Jalasoft.TeamUp.Resumes.API.Tests/PostResumesTests.cs]: Also add unit tests for the service layer

File: .idea/.gitignore [hunk 8]
```diff
@@ -0,0 +1,5 @@
+# Default ignored files
```
Comment (cristian-chavez-jalafund) [inline .idea/.gitignore]: please could you remove this .idea/ files or am I missing something? I think we should not add these files, our main .gitignore is the one at the root of our folder

File: src/api/Jalasoft.TeamUp.Resumes.API/Controllers/PostResume.cs [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    public class PostResume
+    {
+        private readonly IResumesService resumesService;
+
+        public PostResume(IResumesService resumesService)
+        {
+            this.resumesService = resumesService;
+        }
+
+        [FunctionName("resume")]
```
Comment (cristian-chavez-jalafund) [inline src/api/Jalasoft.TeamUp.Resumes.API/Controllers/PostResume.cs]: PostResume

File: src/api/Jalasoft.TeamUp.Resumes.API/Controllers/PostResume.cs [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+        [OpenApiOperation(operationId: "createResume", tags: new[] { "Resumes" })]
+        [OpenApiRequestBody("application/json", typeof(Resume), Description = "JSON request body containing { FirstName, LastName, Email, Phone, Summary}")]
```
Comment (cristian-chavez-jalafund) [inline src/api/Jalasoft.TeamUp.Resumes.API/Controllers/PostResume.cs]: please double-check the request,  so far I saw in the model you'll be posting more attributes than { FirstName, LastName, Email, Phone, Summary}

File: src/api/Jalasoft.TeamUp.Resumes.API/Controllers/PostResume.cs [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+        [OpenApiResponseWithBody(statusCode: HttpStatusCode.OK, contentType: "application/json", bodyType: typeof(Resume), Description = "Successful response")]
```
Comment (cristian-chavez-jalafund) [inline src/api/Jalasoft.TeamUp.Resumes.API/Controllers/PostResume.cs]: please return the proper HttpStatusCode for a POST

File: src/api/Jalasoft.TeamUp.Resumes.API/Controllers/PostResume.cs [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+        public IActionResult CreateResume(
+            [HttpTrigger(AuthorizationLevel.Anonymous, "post", Route = "v1/resumes")] HttpRequest req)
+        {
+            string requestBody = new StreamReader(req.Body).ReadToEnd();
+            var input = JsonConvert.DeserializeObject<Resume>(requestBody);
+            var createResume = this.resumesService.PostResumes(input);
+            return new OkObjectResult(createResume);
```
Comment (cristian-chavez-jalafund) [inline src/api/Jalasoft.TeamUp.Resumes.API/Controllers/PostResume.cs]: please return the proper HttpStatusCode for a POST

**Review by cristian-chavez-jalafund — CHANGES_REQUESTED** [review]

File: src/api/Jalasoft.TeamUp.Resumes.API/Controllers/PostResume.cs (hunk 9 above)
Comment (gustavo-rojas-jalafund) [reply src/api/Jalasoft.TeamUp.Resumes.API/Controllers/PostResume.cs]: Resolved

File: .idea/.gitignore (hunk 8 above)
Comment (gustavo-rojas-jalafund) [reply .idea/.gitignore]: Remove and resolved

File: tests/api/Jalasoft.TeamUp.Resumes.API.Tests/PostResumesTests.cs (hunk 7 above)
Comment (gustavo-rojas-jalafund) [reply tests/api/Jalasoft.TeamUp.Resumes.API.Tests/PostResumesTests.cs]: Done

File: src/dal/Jalasoft.TeamUp.Resumes.DAL/ResumesRepository.cs (hunk 6 above)
Comment (gustavo-rojas-jalafund) [reply src/dal/Jalasoft.TeamUp.Resumes.DAL/ResumesRepository.cs:97]: Resolved

File: src/api/Jalasoft.TeamUp.Resumes.API/Controllers/PostResume.cs (hunk 10 above)
Comment (gustavo-rojas-jalafund) [reply src/api/Jalasoft.TeamUp.Resumes.API/Controllers/PostResume.cs]: Added attribute picture

File: src/api/Jalasoft.TeamUp.Resumes.API/Controllers/PostResume.cs (hunk 12 above)
Comment (gustavo-rojas-jalafund) [reply src/api/Jalasoft.TeamUp.Resumes.API/Controllers/PostResume.cs]: Resolved

File: src/core/Jalasoft.TeamUp.Resumes.Core.Interfaces/IHealthService.cs (hunk 1 above)
Comment (gustavo-rojas-jalafund) [reply src/core/Jalasoft.TeamUp.Resumes.Core.Interfaces/IHealthService.cs]: Health PostHealth() was remove

File: src/dal/Jalasoft.TeamUp.Resumes.DAL/ResumesRepository.cs (hunk 5 above)
Comment (gustavo-rojas-jalafund) [reply src/dal/Jalasoft.TeamUp.Resumes.DAL/ResumesRepository.cs:11]: Resolved

File: src/api/Jalasoft.TeamUp.Resumes.API/Controllers/PostResume.cs (hunk 11 above)
Comment (Raul-Gamarra-jalafund) [reply src/api/Jalasoft.TeamUp.Resumes.API/Controllers/PostResume.cs]: `statusCode: HttpStatusCode.OK`
this status code should be "Created"

File: src/core/Jalasoft.TeamUp.Resumes.Core.Interfaces/IHealthService.cs [hunk 13]
```diff
@@ -5,5 +5,7 @@
     public interface IHealthService
     {
         Health GetHealth();
+
+        // Health PostHealth();
```
Comment (Raul-Gamarra-jalafund) [inline src/core/Jalasoft.TeamUp.Resumes.Core.Interfaces/IHealthService.cs]: please remove, don't push commented code

File: src/dal/Jalasoft.TeamUp.Resumes.DAL.Interfaces/IHealthRepository.cs (hunk 3 above)
Comment (Raul-Gamarra-jalafund) [reply src/dal/Jalasoft.TeamUp.Resumes.DAL.Interfaces/IHealthRepository.cs]: don't forget to delete this

File: src/dal/Jalasoft.TeamUp.Resumes.DAL/HealthRepository.cs (hunk 4 above)
Comment (Raul-Gamarra-jalafund) [inline src/dal/Jalasoft.TeamUp.Resumes.DAL/HealthRepository.cs]: please remove this method as we do not need a health creation

File: src/dal/Jalasoft.TeamUp.Resumes.DAL/ResumesRepository.cs [hunk 14]
```diff
@@ -7,7 +7,8 @@
 
     public class ResumesRepository : IResumesRepository
     {
-        private static readonly Resume[] Resumes = new Resume[]
+        // private static List<Resume> resumes = new List<Resume>
```
Comment (Raul-Gamarra-jalafund) [inline src/dal/Jalasoft.TeamUp.Resumes.DAL/ResumesRepository.cs]: please remove commented code

File: tests/api/Jalasoft.TeamUp.Resumes.API.Tests/PostResumesTests.cs [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+                            NameSkill = "API"
+                        }
+                    },
+                CreationDate = DateTime.Now.AddDays(-10),
+                LastUpdate = DateTime.Now
+            };
+
+            this.mockRepository.Setup(repository => repository.PostResumes(resume)).Returns(new Resume());
+            var result = this.resumeService.PostResumes(resume);
+            Assert.IsType<Resume>(result);
```
Comment (Raul-Gamarra-jalafund) [inline tests/api/Jalasoft.TeamUp.Resumes.API.Tests/PostResumesTests.cs]: you are missing the status code validation in this unit test

File: tests/api/Jalasoft.TeamUp.Resumes.API.Tests/PostResumesTests.cs [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+        }
+    }
+}
```
Comment (Raul-Gamarra-jalafund) [inline tests/api/Jalasoft.TeamUp.Resumes.API.Tests/PostResumesTests.cs]: please add unit tests for the service as well

File: src/core/Jalasoft.TeamUp.Resumes.Core.Interfaces/IHealthService.cs (hunk 13 above)
Comment (gustavo-rojas-jalafund) [reply src/core/Jalasoft.TeamUp.Resumes.Core.Interfaces/IHealthService.cs]: Resolved

File: src/dal/Jalasoft.TeamUp.Resumes.DAL/ResumesRepository.cs (hunk 14 above)
Comment (gustavo-rojas-jalafund) [reply src/dal/Jalasoft.TeamUp.Resumes.DAL/ResumesRepository.cs]: Resolved

File: tests/api/Jalasoft.TeamUp.Resumes.API.Tests/PostResumesTests.cs (hunk 16 above)
Comment (gustavo-rojas-jalafund) [reply tests/api/Jalasoft.TeamUp.Resumes.API.Tests/PostResumesTests.cs]: Done

File: src/dal/Jalasoft.TeamUp.Resumes.DAL.Interfaces/IHealthRepository.cs (hunk 3 above)
Comment (gustavo-rojas-jalafund) [reply src/dal/Jalasoft.TeamUp.Resumes.DAL.Interfaces/IHealthRepository.cs]: Done

File: src/api/Jalasoft.TeamUp.Resumes.API/Controllers/PostResume.cs (hunk 10 above)
Comment (gustavo-rojas-jalafund) [reply src/api/Jalasoft.TeamUp.Resumes.API/Controllers/PostResume.cs]: Resolved

**Review by ela-ocana-jalafund — DISMISSED** [review]

**Review by rodrigo-baldiviezo-jalafund — DISMISSED** [review]

**Review by cristian-chavez-jalafund — DISMISSED** [review]

File: src/dal/Jalasoft.TeamUp.Resumes.DAL/ResumesRepository.cs [hunk 17]
```diff
@@ -82,13 +82,19 @@ public class ResumesRepository : IResumesRepository
...
 
         public IEnumerable<Resume> GetResumes()
         {
-            return Resumes;
+            return resumes;
+        }
+
+        public Resume PostResumes(Resume resume)
+        {
+            resumes = new List<Resume>(resumes) { resume }.ToArray();
```
Comment (cristian-chavez-jalafund) [inline src/dal/Jalasoft.TeamUp.Resumes.DAL/ResumesRepository.cs:96]:
```suggestion
            resumes.Add(resume);
```

**Review by Raul-Gamarra-jalafund — DISMISSED** [review]

File: tests/api/Jalasoft.TeamUp.Resumes.API.Tests/PostResumesApiTests.cs [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+            this.postResume = new PostResume(this.mockResumesService.Object);
+        }
+
+        [Fact]
+        public void PostResume_Returns_CreateResume_Resume()
+        {
+            var request = this.mockHttpContext.Request;
+            this.mockResumesService.Setup(service => service.PostResumes(null)).Returns(new Resume());
+            var response = this.postResume.CreateResume(request);
+            var okObjectResult = Assert.IsType<CreatedResult>(response);
```
Comment (cristian-chavez-jalafund) [inline tests/api/Jalasoft.TeamUp.Resumes.API.Tests/PostResumesApiTests.cs]:
```suggestion
            var createdResult = Assert.IsType<CreatedResult>(response);
```

File: src/dal/Jalasoft.TeamUp.Resumes.DAL/ResumesRepository.cs (hunk 17 above)
Comment (gustavo-rojas-jalafund) [reply src/dal/Jalasoft.TeamUp.Resumes.DAL/ResumesRepository.cs:96]: Sorry Cristian, resumes is an array and not have the Add function

**Review by rodrigo-baldiviezo-jalafund — APPROVED** [review]
LGTM

**Review by Raul-Gamarra-jalafund — APPROVED** [review]

**Review by cristian-chavez-jalafund — APPROVED** [review]

**Pull request merged**